File: frontend/src/types/trial-protocol.ts
export type TrialProtocolItemDraft = {
  ordinal: number;
  name: string;
  required: boolean;
  enabled: boolean;
};

export type TrialProtocolVisitDraft = {
  ordinal: number;
  source_visit_code: string | null;
  name: string;
  window: string | null;
  enabled: boolean;
  items: TrialProtocolItemDraft[];
};

export type TrialProtocolCenterDraft = {
  code: string;
  name: string;
  filing_no: string | null;
  principal_investigator: string | null;
  enabled: boolean;
  confidence?: number | null;
  requires_confirmation?: boolean;
  confirmed?: boolean;
  evidence?: {
    page_no?: number;
    source?: string;
    risk_reasons?: string[];
    lines?: {
      page_no: number;
      text: string;
      source?: string;
      confidence?: number | null;
    }[];
  } | null;
};

export type TrialProtocolDraft = {
  visits: TrialProtocolVisitDraft[];
  centers: TrialProtocolCenterDraft[];
  deactivate_missing: {
    visits: boolean;
    items: boolean;
    centers: boolean;
  };
  parse_meta?: {
    text_source?: string;
    page_count?: number;
    center_count?: number;
    center_risk_count?: number;
    evidence_mode?: string;
  } | null;
};

export type TrialProtocolVersion = {
  id: number;
  project_id: number;
  version_number: number;
  original_name: string;
  file_hash: string;
  file_size: number;
  page_count: number;
  parsing_status: string;
  protocol_no: string | null;
  protocol_version: string | null;
  protocol_date: string | null;
  draft_json: TrialProtocolDraft;
  apply_result_json: Record<string, number> | null;
  uploaded_by: number | null;
  applied_by: number | null;
  uploaded_at: string;
  applied_at: string | null;
  created_at: string;
  updated_at: string;
};

export type TrialProtocolVersionSummary = Omit<
  TrialProtocolVersion,
  "draft_json" | "uploaded_by" | "applied_by" | "created_at" | "updated_at"
>;

export type TrialProtocolApplyResult = {
  version: TrialProtocolVersion;
  result: Record<string, number>;
};

File: backend/app/schemas/trial_protocol.py
from datetime import datetime

from pydantic import BaseModel, ConfigDict, Field


class TrialProtocolItemDraft(BaseModel):
    ordinal: int = Field(ge=1)
    name: str = Field(min_length=1, max_length=150)
    required: bool = True
    enabled: bool = True


class TrialProtocolVisitDraft(BaseModel):
    ordinal: int = Field(ge=1)
    source_visit_code: str | None = Field(default=None, max_length=30)
    name: str = Field(min_length=1, max_length=100)
    window: str | None = Field(default=None, max_length=120)
    enabled: bool = True
    items: list[TrialProtocolItemDraft] = Field(default_factory=list)


class TrialProtocolCenterDraft(BaseModel):
    code: str = Field(min_length=1, max_length=50)
    name: str = Field(min_length=1, max_length=100)
    filing_no: str | None = Field(default=None, max_length=80)
    principal_investigator: str | None = Field(default=None, max_length=100)
    enabled: bool = True
    confidence: float | None = Field(default=None, ge=0, le=1)
    requires_confirmation: bool = False
    confirmed: bool = True
    evidence: dict | None = None


class TrialProtocolDeactivateMissing(BaseModel):
    visits: bool = False
    items: bool = False
    centers: bool = False


class TrialProtocolDraft(BaseModel):
    visits: list[TrialProtocolVisitDraft] = Field(default_factory=list)
    centers: list[TrialProtocolCenterDraft] = Field(default_factory=list)
    deactivate_missing: TrialProtocolDeactivateMissing = Field(
        default_factory=TrialProtocolDeactivateMissing
    )
    parse_meta: dict | None = None


class TrialProtocolVersionRead(BaseModel):
    id: int
    project_id: int
    version_number: int
    original_name: str
    file_hash: str
    file_size: int
    page_count: int
    parsing_status: str
    protocol_no: str | None
    protocol_version: str | None
    protocol_date: str | None
    draft_json: TrialProtocolDraft
    apply_result_json: dict | None
    uploaded_by: int | None
    applied_by: int | None
    uploaded_at: datetime
    applied_at: datetime | None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)


class TrialProtocolVersionSummary(BaseModel):
    id: int
    project_id: int
    version_number: int
    original_name: str
    file_hash: str
    file_size: int
    page_count: int
    parsing_status: str
    protocol_no: str | None
    protocol_version: str | None
    protocol_date: str | None
    apply_result_json: dict | None
    uploaded_at: datetime
    applied_at: datetime | None

    model_config = ConfigDict(from_attributes=True)


class TrialProtocolApplyResult(BaseModel):
    version: TrialProtocolVersionRead
    result: dict
